### PR TITLE
feat: commitMessageLowerCase

### PR DESCRIPTION
**Pull request by kingcrunch**
Semantic-commits states, that only the first letter should be lowercased

## Changes

Change the commit title formatting from "all lowercase" to "first letter lowercase" for semantic-commits.

## Context

Semantic commits states, that only the first letter of the commit message should be lowercase. However, although this is not part of this change, there may be cases, where one needs to ignore this rule, for example if one needs to add JIRA-IDs to the commit title, which are traditionally all uppercase (sigh).

This is a follow-up of the discussion https://github.com/renovatebot/renovate/discussions/20923

See also

- #12062
- #13197
- #16057
- #16239
- #16263

It does (intentionally) _not_ touch the PR title

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by KingCrunch** [comment]
To be honest: Just a wild guess, because not my language (and so on). I am open for recommendations, suggestions, whatever.

**Review by viceice — CHANGES_REQUESTED** [review]
doesn't work as expected, it would only lowercase the first char of the semantic commit type

**Comment by HonkingGoose** [comment]
I'm labeling this PR `breaking` because of this quote from `@rarkins`: [^quote]

> Let's say we remove all commit message lower casing, probably in a major release to flag it's a breaking change. 

[^quote]: https://github.com/renovatebot/renovate/discussions/20931#discussioncomment-5351645

**Comment by rarkins** [comment]
Let's refer back to the discussion until consensus is reached

**Comment by KingCrunch** [comment]
> doesn't work as expected, it would only lowercase the first char of the semantic commit type

@viceice It's replaced now.

**Comment by KingCrunch** [comment]
> `build / test` and `build / lint` fail, but the issues seems to be unrelated to this PR.

It _was_ related. Fixed :)

File: lib/config/types.ts [hunk 1]
```diff
@@ -77,6 +77,7 @@ export interface RenovateSharedConfig {
   automergeSchedule?: string[];
   semanticCommits?: 'auto' | 'enabled' | 'disabled';
   semanticCommitScope?: string | null;
+  commitMessageLowerCase?: string | null;
```
Comment (viceice) [inline lib/config/types.ts:80]: use string constants, like `semanticCommits` above.

File: lib/modules/platform/azure/index.ts [hunk 2]
```diff
@@ -301,11 +301,15 @@ export async function findPr({
     const prs = await getPrList();
 
     prsFiltered = prs.filter(
-      (item) => item.sourceRefName === getNewBranchName(branchName)
+      (item) =>
+        item.sourceRefName?.toLowerCase() ===
+        getNewBranchName(branchName)?.toLowerCase()
```
Comment (viceice) [inline lib/modules/platform/azure/index.ts:306]: i don't think we need this, the new option shouldn't change branch names.

File: lib/modules/platform/bitbucket-server/index.ts [hunk 3]
```diff
@@ -286,8 +286,8 @@ function matchesState(state: string, desiredState: string): boolean {
 const isRelevantPr =
   (branchName: string, prTitle: string | null | undefined, state: string) =>
   (p: Pr): boolean =>
-    p.sourceBranch === branchName &&
-    (!prTitle || p.title === prTitle) &&
+    p.sourceBranch.toLowerCase() === branchName.toLowerCase() &&
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/index.ts:289]: same as above

File: lib/modules/platform/azure/index.ts [hunk 4]
```diff
@@ -301,11 +301,15 @@ export async function findPr({
...
     );
 
     if (prTitle) {
-      prsFiltered = prsFiltered.filter((item) => item.title === prTitle);
+      prsFiltered = prsFiltered.filter(
+        (item) => item.title.toLowerCase() === prTitle.toLowerCase()
```
Comment (viceice) [inline lib/modules/platform/azure/index.ts:311]: use `toUpperCase` as there are chars which doesn't have a lower case equivalent

File: lib/workers/repository/updates/generate.ts [hunk 5]
```diff
@@ -226,7 +223,7 @@ export function generateBranchConfig(
       regEx(/to vv(\d)/),
       'to v$1'
     );
-    if (upgrade.toLowerCase) {
+    if (upgrade.commitMessageLowerCase === 'auto') {
```
Comment (viceice) [inline lib/workers/repository/updates/generate.ts:226]: this is not the same behavior as before, we need to ensure existing behavior

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/platform/azure/index.ts (hunk 4 above)
Comment (KingCrunch) [reply lib/modules/platform/azure/index.ts:311]: No problem. Out of curiosity: What happens with the characters? The languages I know just keep them untouched, which would make the behaviour equivalent.

Comment (viceice) [reply lib/modules/platform/azure/index.ts]: I first saw a lint rule on dotnet, which suggested to use upper case


https://stackoverflow.com/a/14128850

**Comment by KingCrunch** [comment]
The `breaking` flag can be remove, no?

**Comment by HonkingGoose** [comment]
@rarkins is the current PR still breaking? Please keep/remove the `breaking` label as needed. 😉

File: lib/config/types.ts [hunk 6]
```diff
@@ -77,6 +77,7 @@ export interface RenovateSharedConfig {
   automergeSchedule?: string[];
   semanticCommits?: 'auto' | 'enabled' | 'disabled';
   semanticCommitScope?: string | null;
+  commitMessageLowerCase?: 'auto' | 'never';
```
Comment (viceice) [inline lib/config/types.ts:80]: @JamieMagee should we add undefined for typescript?

File: lib/workers/repository/updates/generate.ts [hunk 7]
```diff
@@ -226,7 +226,7 @@ export function generateBranchConfig(
       regEx(/to vv(\d)/),
       'to v$1'
     );
-    if (upgrade.toLowerCase) {
+    if (upgrade.toLowerCase && upgrade.commitMessageLowerCase === 'auto') {
```
Comment (viceice) [inline lib/workers/repository/updates/generate.ts:229]:
```suggestion
    if (upgrade.toLowerCase && upgrade.commitMessageLowerCase !== 'never') {
```

@rarkins should we negate to be more save against missing value?

File: lib/workers/repository/updates/generate.ts [hunk 8]
```diff
@@ -249,7 +249,7 @@ export function generateBranchConfig(
         );
         throw new Error(CONFIG_SECRETS_EXPOSED);
       }
-      if (upgrade.toLowerCase) {
+      if (upgrade.toLowerCase && upgrade.commitMessageLowerCase === 'auto') {
```
Comment (viceice) [inline lib/workers/repository/updates/generate.ts:252]:
```suggestion
      if (upgrade.toLowerCase && upgrade.commitMessageLowerCase !== 'never') {
```

**Comment by kingcrunch** [comment]
Is there is really no interest in fixing this bug?

Btw the PR is _still_ no breaking change. It was, but I changed it.

So I don't know, why this is still postponed to the next major release.

At the end for me it was much easier to change the company rules, than fixing a bug here...

**Comment by rarkins** [comment]
There are currently code changes requested

**Comment by kingcrunch** [comment]
Hi @rarkins 

Maybe there is a misunderstanding, but I cannot see any request for change. I can see two question. One for you

> @rarkins should we negate to be more save against missing value?
https://github.com/renovatebot/renovate/pull/20930/files#r1154877720

and one for @JamieMagee 

> @JamieMagee should we add undefined for typescript?
https://github.com/renovatebot/renovate/pull/20930/files#r1154874172

So this seems to be still under discussion and it doesn't make sense for me to propose changes, that are still discussed.

I don't know your workflows, but to me it seems nobody cares...

**Comment by rarkins** [comment]
Please check for the currently 2 unresolved comments in https://github.com/renovatebot/renovate/pull/20930/files

**Review by rarkins — CHANGES_REQUESTED** [review]

**Review by rarkins — DISMISSED** [review]

File: docs/usage/configuration-options.md [hunk 9]
```diff
@@ -3310,6 +3310,11 @@ If configured to `enabled`, then the `semanticCommitScope` and `semanticCommitTy
 
 Renovate autodetects if your repository is already using semantic commits or not and follows suit, so you only need to configure this if you wish to _override_ Renovate's autodetected setting.
 
+## commitMessageLowerCase
```
Comment (viceice) [inline docs/usage/configuration-options.md:3313]: wrong place, sort config options alphabetical

**Review by viceice — COMMENTED** [review]
markdown config needs sorted, otherwise LGTM

**Review by viceice — COMMENTED** [review]
markdown config needs sorted, otherwise LGTM

**Review by rarkins — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**